pkg/common/workapplier/workapplier.go

```diff
@@ -0,0 +1,123 @@
+package workapplier
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	workv1client "open-cluster-management.io/api/client/work/clientset/versioned"
+	worklister "open-cluster-management.io/api/client/work/listers/work/v1"
+	workapiv1 "open-cluster-management.io/api/work/v1"
+)
+
+type WorkApplier struct {
+	Client workv1client.Interface
```

Comment:
Instead of having client and lister defined here, you might want to define 

```go
getWork func(ctx context, namespace, name string) (work, error)
deleteWork func(ctx context, namespace, name string) error
updatedWork ...
createWork ...
```

and a 
```go
NewWorkApplierWithTypedClient(workClient workv1client.Interface, workLister worklister.ManifestWorkLister)
```

This gives controller-runtime a way also to define its own applier.

Reply:
done